ivy/functional/frontends/numpy/ndarray/ndarray.py

```diff
@@ -488,3 +488,11 @@ def __iter__(self):
 
     def __mod__(self, value, /):
         return np_frontend.mod(self, value, out=self)
+    
+    def ptp(self, *, axis=None, out=None, keepdims=False):
+        if isinstance(self._ivy_array, int):
+            return ivy.flatten(self._ivy_array, order="C")
+        else:
+            xmax = ivy.max(self, axis=axis, out=out, keepdims=keepdims)
+            xmin = ivy.min(self, axis=axis, out=out, keepdims=keepdims)
```

Comment:
I think the error would go away if you replaced this with:
```
xmax = self.max(axis=axis, out=out, keepdims=keepdims)
xmin = self.min(axis=axis, out=out, keepdims=keepdims)
```
Ivy functions cannot handle ndarrays as input. They expect either ivy arrays or native arrays.

Reply:
ivy_tests/test_ivy/test_frontends/test_numpy/test_ndarray/test_ndarray.py:2914 (test_numpy_instance_ptp[cpu-ivy.functional.backends.numpy-False-False])
args = (ivy.array(0),), kwargs = {'axis': 1, 'keepdims': False, 'out': None}

    @functools.wraps(fn)
    def _handle_exceptions(*args, **kwargs):
        """
        Catch all exceptions and raise them in IvyException.
    
        Parameters
        ----------
        args
            The arguments to be passed to the function.
    
        kwargs
            The keyword arguments to be passed to the function.
    
        Returns
        -------
            The return of the function, or raise IvyException if error is thrown.
        """
        try:
>           return fn(*args, **kwargs)

../../../../../ivy/utils/exceptions.py:122: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (ivy.array(0),), kwargs = {'axis': 1, 'keepdims': False, 'out': None}
cont_fn = <function handle_nestable.<locals>._handle_nestable.<locals>.<lambda> at 0x29382b160>

    @functools.wraps(fn)
    def _handle_nestable(*args, **kwargs):
        """
        Call `fn` with the *nestable* property of the function correctly handled. This
        means mapping the function to the container leaves if any containers are passed
        in the input.
    
        Parameters
        ----------
        args
            The arguments to be passed to the function.
    
        kwargs
            The keyword arguments to be passed to the function.
    
        Returns
        -------
            The return of the function, with the nestable property handled correctly.
        """
        # if any of the arguments or keyword arguments passed to the function contains
        # a container, get the container's version of the function and call it using
        # the passed arguments.
        if hasattr(ivy.Container, "_static_" + fn_name):
            cont_fn = getattr(ivy.Container, "_static_" + fn_name)
        else:
            cont_fn = lambda *args, **kwargs: ivy.Container.cont_multi_map_in_function(
                fn, *args, **kwargs
            )
        if ivy.get_nestable_mode() and (
            ivy.nested_any(args, ivy.is_ivy_container, check_nests=True)
            or ivy.nested_any(kwargs, ivy.is_ivy_container, check_nests=True)
        ):
            return cont_fn(*args, **kwargs)
    
        # if the passed arguments does not contain a container, the function using
        # the passed arguments, returning an ivy or a native array.
>       return fn(*args, **kwargs)

../../../../../ivy/func_wrapper.py:752: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = [ivy.array(0)], kwargs = {'axis': 1, 'keepdims': False, 'out': None}
num_args = 1
type_hints = mappingproxy(OrderedDict([('x', <Parameter "x: Union[ivy.data_classes.array.array.Array, ivy.NativeArray]">), ('axis',...r "keepdims: bool = False">), ('out', <Parameter "out: Union[ivy.data_classes.array.array.Array, NoneType] = None">)]))
parameters = ['x', 'axis', 'keepdims', 'out']
annotations = [typing.Union[ivy.data_classes.array.array.Array, ivy.NativeArray], typing.Union[int, typing.Sequence[int], NoneType], <class 'bool'>, typing.Union[ivy.data_classes.array.array.Array, NoneType]]
i = 0
annotation = typing.Union[ivy.data_classes.array.array.Array, ivy.NativeArray]
parameter = 'x', arg = ivy.array(0)

    @functools.wraps(fn)
    def _handle_array_like_without_promotion(*args, **kwargs):
        args = list(args)
        num_args = len(args)
        try:
            type_hints = inspect.signature(fn).parameters
        except (TypeError, ValueError):
            return fn(*args, **kwargs)
        parameters = list(type_hints.keys())
        annotations = [param.annotation for param in type_hints.values()]
    
        for i, (annotation, parameter, arg) in enumerate(
            zip(annotations, parameters, args)
        ):
            annotation_str = str(annotation)
            if (
                ("rray" in annotation_str or "Tensor" in annotation_str)
                and parameter != "out"
                and all(
                    sq not in annotation_str
                    for sq in ["Sequence", "List", "Tuple", "float", "int", "bool"]
                )
            ):
                if i < num_args:
                    # Fix for ellipsis, slices for numpy's __getitem__
                    # No need to try and convert them into arrays
                    # since asarray throws unpredictable bugs
                    if _check_in_nested_sequence(arg, value=Ellipsis, _type=slice):
                        continue
                    if not ivy.is_array(arg):
                        args[i] = ivy.array(arg)
                elif parameters in kwargs:
                    kwarg = kwargs[parameter]
                    if not ivy.is_array(kwarg):
                        kwargs[parameter] = ivy.array(kwarg)
    
>       return fn(*args, **kwargs)

../../../../../ivy/func_wrapper.py:244: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

out = None, args = (ivy.array(0),), kwargs = {'axis': 1, 'keepdims': False}

    @functools.wraps(fn)
    def _handle_out_argument(*args, out=None, **kwargs):
        """
        Call `fn` with the `out` argument handled correctly for performing an inplace
        update.
    
        Parameters
        ----------
        args
            The arguments to be passed to the function.
    
        out
            The array to write the result to.
    
        kwargs
            The keyword arguments to be passed to the function.
    
        Returns
        -------
            The return of the function, with `out` handled correctly for
            inplace updates.
        """
        if out is None or handle_out_in_ivy:
>           return fn(*args, out=out, **kwargs)

../../../../../ivy/func_wrapper.py:646: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (ivy.array(0),), kwargs = {'axis': 1, 'keepdims': False}, has_out = True
out = None, new_args = (array(0, dtype=int32),)
new_kwargs = {'axis': 1, 'keepdims': False, 'out': None}

    @functools.wraps(fn)
    def _inputs_to_native_arrays(*args, **kwargs):
        """
        Convert all `ivy.Array` instances in both the positional and keyword arguments
        into `ivy.NativeArray` instances, and then calls the function with the updated
        arguments.
    
        Parameters
        ----------
        args
            The arguments to be passed to the function.
    
        kwargs
            The keyword arguments to be passed to the function.
    
        Returns
        -------
            The return of the function, with native arrays passed in the arguments.
        """
        if not ivy.get_array_mode():
            return fn(*args, **kwargs)
        # check if kwargs contains an out argument, and if so, remove it
        has_out = False
        out = None
        if "out" in kwargs:
            out = kwargs["out"]
            del kwargs["out"]
            has_out = True
        # convert all arrays in the inputs to ivy.NativeArray instances
        new_args, new_kwargs = ivy.args_to_native(*args, **kwargs)
        # add the original out argument back to the keyword arguments
        if has_out:
            new_kwargs["out"] = out
>       return fn(*new_args, **new_kwargs)

../../../../../ivy/func_wrapper.py:284: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (array(0, dtype=int32),)
kwargs = {'axis': 1, 'keepdims': False, 'out': None}

    @functools.wraps(fn)
    def _outputs_to_ivy_arrays(*args, **kwargs):
        """
        Call the function, and then converts all `ivy.NativeArray` instances in the
        function return into `ivy.Array` instances.
    
        Parameters
        ----------
        args
            The arguments to be passed to the function.
    
        kwargs
            The keyword arguments to be passed to the function.
    
        Returns
        -------
            The return of the function, with native arrays as ivy arrays.
        """
        # call unmodified function
>       ret = fn(*args, **kwargs)

../../../../../ivy/func_wrapper.py:376: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (array(0, dtype=int32),)
kwargs = {'axis': 1, 'keepdims': False, 'out': None}, overloaded_types = []
overloaded_args = [], arg = None, success = False, value = None

    @functools.wraps(func)
    def _handle_array_function(*args, **kwargs):
        overloaded_types = []
        overloaded_args = []
    
        for arg in args + tuple(kwargs.values()):
            if ivy.exists(arg) and (
                not isinstance(arg, ivy.Container)
                and hasattr(arg, "__ivy_array_function__")
            ):
                if type(arg) not in overloaded_types:
                    overloaded_types.append(type(arg))
                    if (
                        arg.__ivy_array_function__
                        is not ivy.Array.__ivy_array_function__
                        and not isinstance(arg, (ivy.Array, ivy.NativeArray))
                    ):
                        index = len(overloaded_args)
                        for i, old_arg in enumerate(overloaded_args):
                            if issubclass(type(arg), type(old_arg)):
                                index = i
                                break
                        overloaded_args.insert(index, arg)
            if ivy.exists(arg) and isinstance(arg, ivy.Container):
                arg = ivy.Container.cont_flatten_key_chains(arg)
                indices = ivy.nested_argwhere(
                    arg, lambda x: hasattr(x, "__ivy_array_function__")
                )
                for a in indices:
                    if type(getattr(arg, a[0])) not in overloaded_types:
                        overloaded_types.append(type(getattr(arg, a[0])))
    
                        if getattr(
                            arg, a[0]
                        ).__ivy_array_function__ is not ivy.Array.__ivy_array_function__ and not isinstance(  # noqa: E501
                            getattr(arg, a[0]), (ivy.Array, ivy.NativeArray)
                        ):
                            index = len(overloaded_args)
                            for i, old_arg in enumerate(overloaded_args):
                                if issubclass(type(getattr(arg, a[0])), type(old_arg)):
                                    index = i
                                    break
                            overloaded_args.insert(index, arg)
    
        success, value = try_array_function_override(
            ivy.__dict__[func.__name__], overloaded_args, overloaded_types, args, kwargs
        )
        if success:
            return value
>       return func(*args, **kwargs)

../../../../../ivy/func_wrapper.py:201: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

x = array(0, dtype=int32)

    def max(
        x: np.ndarray,
        /,
        *,
        axis: Optional[Union[int, Sequence[int]]] = None,
        keepdims: bool = False,
        out: Optional[np.ndarray] = None,
    ) -> np.ndarray:
        axis = tuple(axis) if isinstance(axis, list) else axis
>       return np.asarray(np.amax(a=x, axis=axis, keepdims=keepdims, out=out))

../../../../../ivy/functional/backends/numpy/statistical.py:40: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = ()
kwargs = {'a': array(0, dtype=int32), 'axis': 1, 'keepdims': False, 'out': None}
relevant_args = (array(0, dtype=int32), None)

>   ???

<__array_function__ internals>:200: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = array(0, dtype=int32), axis = 1, out = None, keepdims = False
initial = <no value>, where = <no value>

    @array_function_dispatch(_amax_dispatcher)
    def amax(a, axis=None, out=None, keepdims=np._NoValue, initial=np._NoValue,
             where=np._NoValue):
        """
        Return the maximum of an array or maximum along an axis.
    
        Parameters
        ----------
        a : array_like
            Input data.
        axis : None or int or tuple of ints, optional
            Axis or axes along which to operate.  By default, flattened input is
            used.
    
            .. versionadded:: 1.7.0
    
            If this is a tuple of ints, the maximum is selected over multiple axes,
            instead of a single axis or all the axes as before.
        out : ndarray, optional
            Alternative output array in which to place the result.  Must
            be of the same shape and buffer length as the expected output.
            See :ref:`ufuncs-output-type` for more details.
    
        keepdims : bool, optional
            If this is set to True, the axes which are reduced are left
            in the result as dimensions with size one. With this option,
            the result will broadcast correctly against the input array.
    
            If the default value is passed, then `keepdims` will not be
            passed through to the `amax` method of sub-classes of
            `ndarray`, however any non-default value will be.  If the
            sub-class' method does not implement `keepdims` any
            exceptions will be raised.
    
        initial : scalar, optional
            The minimum value of an output element. Must be present to allow
            computation on empty slice. See `~numpy.ufunc.reduce` for details.
    
            .. versionadded:: 1.15.0
    
        where : array_like of bool, optional
            Elements to compare for the maximum. See `~numpy.ufunc.reduce`
            for details.
    
            .. versionadded:: 1.17.0
    
        Returns
        -------
        amax : ndarray or scalar
            Maximum of `a`. If `axis` is None, the result is a scalar value.
            If `axis` is an int, the result is an array of dimension
            ``a.ndim - 1``. If `axis` is a tuple, the result is an array of
            dimension ``a.ndim - len(axis)``.
    
        See Also
        --------
        amin :
            The minimum value of an array along a given axis, propagating any NaNs.
        nanmax :
            The maximum value of an array along a given axis, ignoring any NaNs.
        maximum :
            Element-wise maximum of two arrays, propagating any NaNs.
        fmax :
            Element-wise maximum of two arrays, ignoring any NaNs.
        argmax :
            Return the indices of the maximum values.
    
        nanmin, minimum, fmin
    
        Notes
        -----
        NaN values are propagated, that is if at least one item is NaN, the
        corresponding max value will be NaN as well. To ignore NaN values
        (MATLAB behavior), please use nanmax.
    
        Don't use `amax` for element-wise comparison of 2 arrays; when
        ``a.shape[0]`` is 2, ``maximum(a[0], a[1])`` is faster than
        ``amax(a, axis=0)``.
    
        Examples
        --------
        >>> a = np.arange(4).reshape((2,2))
        >>> a
        array([[0, 1],
               [2, 3]])
        >>> np.amax(a)           # Maximum of the flattened array
        3
        >>> np.amax(a, axis=0)   # Maxima along the first axis
        array([2, 3])
        >>> np.amax(a, axis=1)   # Maxima along the second axis
        array([1, 3])
        >>> np.amax(a, where=[False, True], initial=-1, axis=0)
        array([-1,  3])
        >>> b = np.arange(5, dtype=float)
        >>> b[2] = np.NaN
        >>> np.amax(b)
        nan
        >>> np.amax(b, where=~np.isnan(b), initial=-1)
        4.0
        >>> np.nanmax(b)
        4.0
    
        You can use an initial value to compute the maximum of an empty slice, or
        to initialize it to a different value:
    
        >>> np.amax([[-50], [10]], axis=-1, initial=0)
        array([ 0, 10])
    
        Notice that the initial value is used as one of the elements for which the
        maximum is determined, unlike for the default argument Python's max
        function, which is only used for empty iterables.
    
        >>> np.amax([5], initial=6)
        6
        >>> max([5], default=6)
        5
        """
>       return _wrapreduction(a, np.maximum, 'max', axis, None, out,
                              keepdims=keepdims, initial=initial, where=where)

../../../../../../miniconda3/envs/ivy_dev/lib/python3.8/site-packages/numpy/core/fromnumeric.py:2820: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = array(0, dtype=int32), ufunc = <ufunc 'maximum'>, method = 'max', axis = 1
dtype = None, out = None
kwargs = {'initial': <no value>, 'keepdims': False, 'where': <no value>}
passkwargs = {'keepdims': False}

    def _wrapreduction(obj, ufunc, method, axis, dtype, out, **kwargs):
        passkwargs = {k: v for k, v in kwargs.items()
                      if v is not np._NoValue}
    
        if type(obj) is not mu.ndarray:
            try:
                reduction = getattr(obj, method)
            except AttributeError:
                pass
            else:
                # This branch is needed for reductions like any which don't
                # support a dtype.
                if dtype is not None:
                    return reduction(axis=axis, dtype=dtype, out=out, **passkwargs)
                else:
                    return reduction(axis=axis, out=out, **passkwargs)
    
>       return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
E       numpy.AxisError: axis 1 is out of bounds for array of dimension 0

../../../../../../miniconda3/envs/ivy_dev/lib/python3.8/site-packages/numpy/core/fromnumeric.py:86: AxisError

During handling of the above exception, another exception occurred:

frontend = 'numpy', on_device = 'cpu'

    @handle_frontend_method(
>       class_tree=CLASS_TREE,
        init_tree="numpy.array",
        method_name="ptp",
        dtype_x_axis=helpers.dtype_values_axis(
            available_dtypes=helpers.get_dtypes("numeric"),
            min_axis=None,
            max_axis=None,
            allow_neg_axes=True,
            min_axes_size=1,
            max_axes_size=None,
        ),
    )

test_ndarray.py:2916: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_ndarray.py:2937: in test_numpy_instance_ptp
    helpers.test_frontend_method(
../../../helpers/function_testing.py:1659: in test_frontend_method
    ret, ret_np_flat = get_ret_and_flattened_np_array(
../../../helpers/function_testing.py:1983: in get_ret_and_flattened_np_array
    ret = fn(*args, **kwargs)
../../../../../ivy/functional/frontends/numpy/ndarray/ndarray.py:496: in ptp
    xmax = self.max(axis=axis, out=out, keepdims=keepdims)
../../../../../ivy/functional/frontends/numpy/ndarray/ndarray.py:164: in max
    return np_frontend.amax(
../../../../../ivy/functional/frontends/numpy/func_wrapper.py:514: in _handle_numpy_out
    return fn(*args, **kwargs)
../../../../../ivy/functional/frontends/numpy/func_wrapper.py:396: in _outputs_to_numpy_arrays
    ret = fn(*args, **kwargs)
../../../../../ivy/functional/frontends/numpy/func_wrapper.py:352: in _inputs_to_ivy_arrays_np
    return fn(*ivy_args, **ivy_kwargs)
../../../../../ivy/functional/frontends/numpy/func_wrapper.py:453: in _from_zero_dim_arrays_to_scalar
    ret = fn(*args, **kwargs)
../../../../../ivy/functional/frontends/numpy/mathematical_functions/extrema_finding.py:113: in amax
    res = ivy.max(a, axis=axis, keepdims=keepdims, out=out)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (ivy.array(0),), kwargs = {'axis': 1, 'keepdims': False, 'out': None}

    @functools.wraps(fn)
    def _handle_exceptions(*args, **kwargs):
        """
        Catch all exceptions and raise them in IvyException.
    
        Parameters
        ----------
        args
            The arguments to be passed to the function.
    
        kwargs
            The keyword arguments to be passed to the function.
    
        Returns
        -------
            The return of the function, or raise IvyException if error is thrown.
        """
        try:
            return fn(*args, **kwargs)
        # Not to rethrow as IvyBackendException
        except IvyNotImplementedException as e:
            raise e
        except (IndexError, ValueError, AttributeError) as e:
            _print_traceback_history()
>           raise ivy.utils.exceptions.IvyError(fn.__name__, str(e))
E           ivy.utils.exceptions.IvyError: numpy: max: axis 1 is out of bounds for array of dimension 0

../../../../../ivy/utils/exceptions.py:128: IvyError